controllers/ratelimit_controller.go

```diff
@@ -18,36 +18,185 @@ package controllers
+	serviceUrl := "http://" + limitador.ServiceName + "." + namespace + ".svc.cluster.local:" +
+		strconv.Itoa(limitador.ServiceHTTPPort)
+
+	return url.Parse(serviceUrl)
+}
+
 // RateLimitReconciler reconciles a RateLimit object
 type RateLimitReconciler struct {
 	client.Client
-	Log    logr.Logger
-	Scheme *runtime.Scheme
+	Log                logr.Logger
+	Scheme             *runtime.Scheme
+	limitadorDiscovery LimitadorServiceDiscovery
+}
+
+func NewRateLimitReconciler(kubeClient client.Client, logger logr.Logger, scheme *runtime.Scheme) RateLimitReconciler {
+	limitadorServiceDiscovery := defaultLimitadorServiceDiscovery{}
+
+	return RateLimitReconciler{
+		Client:             kubeClient,
+		Log:                logger,
+		Scheme:             scheme,
+		limitadorDiscovery: &limitadorServiceDiscovery,
+	}
 }
 
 // +kubebuilder:rbac:groups=limitador.3scale.net,resources=ratelimits,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=limitador.3scale.net,resources=ratelimits/status,verbs=get;update;patch
 
 func (r *RateLimitReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	_ = context.Background()
-	_ = r.Log.WithValues("ratelimit", req.NamespacedName)
+	reqLogger := r.Log.WithValues("ratelimit", req.NamespacedName)
+
+	limit := &limitadorv1alpha1.RateLimit{}
+	if err := r.Get(context.TODO(), req.NamespacedName, limit); err != nil {
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		reqLogger.Error(err, "Failed to get RateLimit object.")
+		return ctrl.Result{}, err
+	}
+
+	isLimitMarkedToBeDeleted := limit.GetDeletionTimestamp() != nil
+	if isLimitMarkedToBeDeleted {
+		if contains(limit.GetFinalizers(), rateLimitFinalizer) {
+			if err := r.finalizeRateLimit(limit); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			// Remove finalizer. Once all finalizers have been removed, the
+			// object will be deleted.
+			controllerutil.RemoveFinalizer(limit, rateLimitFinalizer)
+			if err := r.Update(context.TODO(), limit); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
 
-	// your logic here
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.ensureFinalizerIsAdded(limit, reqLogger); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.createLimitInLimitador(limit); err != nil {
+		reqLogger.Error(err, "Failed to create rate limit in Limitador.")
+		return ctrl.Result{}, err
+	}
 
 	return ctrl.Result{}, nil
 }
 
 func (r *RateLimitReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&limitadorv1alpha1.RateLimit{}).
+		WithEventFilter(r.updateLimitPredicate()).
 		Complete(r)
 }
+
+// This should be temporary. This is not how a filter should be used. However,
+// with the current Limitador API, when updating a limit, we need both the
+// current and the previous version. After updating the Limitador API to work
+// with IDs, this won't be needed.
```

Comment:
I see. When working with ID's, you will not need to delete and create. It would be enough updating the object in the limitador service

Reply:
Exactly :+1:

Reply:
However this is very limiting. For instance when you want to have CR `status` and manage it to represent rate limits status (available, not available, some error ocurred....). Just food for thought

Reply:
Yes you're right. This needs to be changed. I did it like this only because I wanted to have a first version of the operator that works with the current release of Limitador. Once I improve the Limitador API, I'll come back to this :+1: